src/lj_jitlog.c

```diff
@@ -26,6 +26,32 @@ typedef struct JITLogState {
 #define usr2ctx(usrcontext)  ((JITLogState *)(((char *)usrcontext) - offsetof(JITLogState, user)))
 #define ctx2usr(context)  (&(context)->user)
 
+#if LJ_HASJIT
+
+static void jitlog_exit(JITLogState *context, VMEventData_TExit *exitState)
+{
+  jit_State *J = G2J(context->g);
+  if (J->parent < (1 << 14) && J->exitno < (1 << 10)) {
+    log_traceexit(context->g, exitState->gcexit, J->parent, J->exitno);
+  } else {
+    log_traceexit_large(context->g, exitState->gcexit, J->parent, J->exitno);
+  }
+}
+
+static void jitlog_traceflush(JITLogState *context, FlushReason reason)
+{
+  jit_State *J = G2J(context->g);
+  log_alltraceflush(context->g, reason, J->param[JIT_P_maxtrace], J->param[JIT_P_maxmcode] << 10);
+}
+
+#endif
+
+static void jitlog_gcstate(JITLogState *context, int newstate)
+{
+  global_State *g = context->g;
+  log_gcstate(g, newstate, g->gc.state, g->gc.total, g->strnum);
```

Comment:
it seems that these functions are an interface between a layer that uses `JITLogState` and another (yet to be defined?) that doesn't.  makes me wonder if this means that it would be better to either make `JITLogState` a more "visible" structure, or just pass `g` around.   designing APIs is hard...

Reply:
The API was designed to be usable while keeping JITLogState internal but it look like only one maybe two log_ functions will be used externally. The first design didn't even have state parameter it just used a static SBuf but that wasn't very friendly to running more then one VM in a process.

Reply:
I'm not sure this really answers Javier's question: can you maybe take another look at this question and see if you can concretely answer it? Thanks!

Reply:
well i tried to explain the current design and I want to keep the jitlog state internal.